src/commonMain/kotlin/app/cash/turbine/channel.kt

```diff
@@ -64,16 +77,51 @@ public fun <T> ReceiveChannel<T>.expectNoEvents() {
  *
  * This function will always return a terminal event on a closed [ReceiveChannel].
  */
-public suspend fun <T> ReceiveChannel<T>.awaitEvent(): Event<T> =
-  try {
-    Event.Item(receive())
+@OptIn(ExperimentalCoroutinesApi::class)
+public suspend fun <T> ReceiveChannel<T>.awaitEvent(): Event<T> {
+  val timeout = contextTimeout()
+  val withTimeoutFn: suspend (Duration, suspend CoroutineScope.()->Event<T>)->Event<T> =
+    if (coroutineContext[TestCoroutineScheduler] != null) {
+      // withTimeout uses virtual time, which will hang.
+      ::withWallclockTimeout
+    } else {
+      ::withTimeout
+    }
+  return try {
+    withTimeoutFn(timeout) {
+      val item = receive()
+      Event.Item(item)
+    }
+  } catch (e: TimeoutCancellationException) {
+    throw AssertionError("No value produced in ${timeout}")
+  } catch (e: TurbineTimeoutCancellationException) {
+    throw AssertionError("No value produced in ${timeout}")
   } catch (e: CancellationException) {
     throw e
   } catch (e: ClosedReceiveChannelException) {
     Event.Complete
   } catch (e: Exception) {
     Event.Error(e)
   }
+}
+
+private suspend fun <T> withWallclockTimeout(timeout: Duration, block: suspend CoroutineScope.() -> T): T
+  = coroutineScope {
+    val blockJob = async(start = CoroutineStart.UNDISPATCHED) {
+      block()
+    }
+    val timeoutJob = launch(Dispatchers.Default) {
+      delay(timeout)
+      blockJob.cancel(TurbineTimeoutCancellationException("Timed out waiting for $timeout"))
+    }
+    val result = blockJob.await()
+    timeoutJob.cancel()
+    result
```

Comment:
Could also consider a `select`, although I'm not sure whether it results in clearer/less code.

Probably looks like:
```kotlin
val blockJob = async(start = UNDISPATCHED, block)
val timeoutJob = launch(Dispatchers.Default) { delay(timeout) }
select {
  blockJob.onReceive { result ->
    timeoutJob.cancel()
    result
  }
  timeoutJob.onReceive {
    blockJob.cancel()
    throw TurbineTimeoutCancellationException(..)
  }
}
```

We really want the race API: https://github.com/Kotlin/kotlinx.coroutines/issues/2867

Reply:
Agreed...`blockJob.cancel(TurbineTimeoutCancellationException("Timed out waiting for $timeout"))` is a little weird. Code that still reads clearly even if you don't know about coroutine scope exception propagation is better, so I'll go with `select`